### ISSUE TITLE
Operations randomly failing on multiple reconnects

### ISSUE THREAD

**Issue by dsanghan** [MailCore/mailcore2#1358]
I'm using Reachability in my app along with OAuth2 to maintain session connectivity. If we detect that the internet is connected, the keychain item is used along with gtm-oauth2 to validate the tokens and perform a no-op. If successful, the session is declared online and the overall manager allows operations to be triggered on that session. If we detect a disconnection, the IMAP/SMTP sessions both cancelAllOperations, declare themselves offline to the manager and attempt a disconnectOperation on the IMAP. For the most part, this seems to be working well.

I've noticed that there a few edge cases, where operations do fail with a higher frequency when the duration that the app is open tends to be very large (> 6-7 hours). The operations usually fail with a couldn't find a stable connection to the server error, even though other operations on that sessions dispatch queue succeed before and after the failed operation. The issue disappears on a hard-quit/relaunch. Am I missing any steps in the connect/disconnect flow?

**Comment by dinhviethoa** [comment]
Once you disconnect, could you re-create a new IMAPSession / SMTPSession?

**Comment by dsanghan** [comment]
I am doing that. At the end of the disconnect the current imap/smtp sessions are nilled out. They are recreated when we connect again - sorry if I wasn't clear about that above.

**Comment by dinhviethoa** [comment]
Could you investigate further and let me know?

**Comment by lksnmnn** [comment]
Any updates on this issue @dsanghan ?

**Comment by dsanghan** [comment]
Yes we solved it by adding a few additional steps.

1. For OAuth2, we weren't careful enough about the refresh times for the tokens, so for very long sessions, we need to periodically clean those out and request new ones.
2. For non-OAuth2, we forced full reconnects on temporary connection loses / three successive auth failures
3. For SMTP, checkAccountOperationWithFrom seems to choke out on initial connects so once smtp servers are validated, we just create a new session and use that.
4. Some auto-detection on smtp servers is not correct. We had to figure out when to use SASLLogin instead of Clear.
5. FetchMessagesOperations return inconsistent emails when asking for only UIDs on specific IMAP servers. We had to work around that as well.

Hope that helps. I'll also close this out.
